mltensor-explainer.md

```diff
@@ -0,0 +1,423 @@
+
+- Two fences are created:
+  1. a "start access" fence which is to be signaled by WebNN and waited on by WebGPU. A data copy may be required alongside the signaling of this fence
+  2. an "end access" fence which is to be signaled by WebGPU and waited on by WebNN. A data copy may be required alongside the signaling of this fence
+- The `GPUDevice` enqueues a command to its `GPUQueue` to wait for the "start access" fence to be signaled
+- WebNN will signal the "start access" fence after the completion of all currently-enqueued operations that use the `MLTensor` which is to be imported (this is very similar to how [`GPUBuffer.mapAsync()`](https://www.w3.org/TR/webgpu/#dom-gpubuffer-mapasync) works)
+- Until the "end access" fence is signaled:
+  - The `GPUDevice` has exclusive, read/write access to the imported buffer
+  - All WebNN work involving the imported `MLTensor` is blocked
+- When the `GPUBuffer` is destroyed, the "end access" fence is signaled and the `MLTensor` may be used again by WebNN
+
+### `compute()` vs. `dispatch()`
+
+`compute()` will be deprecated and removed in favor of `dispatch()`.
+
+It's possible `compute()` may have a performance advantage on some platforms for some use cases, such as in one-shot inference (where the benefits of buffer re-use are not relevant) with small inputs/outputs on CPU (where the overhead of task queueing may outweigh the benefits of parallelization). At this time, we do not expect this performance impact to be substantial enough to justify providing two mostly-identical methods for executing an `MLGraph`.
+
+### Open Questions
+
+- How will errors be surfaced? Do we need a concept similar to [WebGPU's error scopes](https://www.w3.org/TR/webgpu/#error-scopes), or is [returning errors via a promise for select operations sufficient](https://github.com/webmachinelearning/webnn/issues/697#issuecomment-2195656878)? See [#477](https://github.com/webmachinelearning/webnn/issues/477)
+- On non-UMA systems, does the user agent have enough information to appropriately allocate an `MLTensor` if an `MLDeviceType` is not used for creating an `MLContext`? See [#350](https://github.com/webmachinelearning/webnn/issues/350) and [#749](https://github.com/webmachinelearning/webnn/issues/749)
+- Should the `dispatch()` method be a part of the `MLGraph` interface rather than `MLContext`? Should `readBuffer()` and `writeBuffer()` exist on an `MLTensor`? See [#697](https://github.com/webmachinelearning/webnn/issues/697).
+- If an `MLContext` is not created from a `GPUDevice`, does there need to be some mechanism - above and beyond the `MLTensorUsage.WEBGPU_INTEROP` flag - for identifying the specific `GPUDevice` with which interop is desired?
+- What are the usage flags of a `GPUBuffer` created from an `MLTensor`?
```

Comment:
`GPUBufferUsage` flags will depend on how the MLTensor was created initially. When I implemented buffer import for D3D12 in Dawn, the only unforeseen restriction was that we couldn’t use `GPUBufferUsage::Uniform` because we couldn’t guarantee the resource was 256 aligned at creation [1]. I don’t know if further cross-platform restrictions exist.

[1] https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/src/dawn/native/d3d12/SharedBufferMemoryD3D12.cpp;l=92

Reply:
> `GPUBufferUsage` flags will depend on how the MLTensor was created initially.

Do you care to elaborate? :)

Reply:
So first - a more formal investigation should be done to solidify all the restrictions on other platforms. I was just making the comment because I have the unique experience of implementing this on D3D12 already.

For D3D12, the only usage restriction I found that differed from a normal GPUBuffer was `UNIFORM` - this was because at the time we couldn't be sure the D3D12Resource would be 256-aligned at creation.

Other than `UNIFORM`, there's not a technical reason you couldn't do interop with any other usages as long as the resource is allocated appropriately. The [code I pointed to](https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/src/dawn/native/d3d12/SharedBufferMemoryD3D12.cpp;l=92) looks at the D3D12_RESOURCE_DESC of the imported resource and automatically assigns all the possible usages based on how it was allocated.

This is kind of why I'm interested in using GPUBuffer as the imported type, rather than a new one. Attributes of a buffer allocated in WebNN could carry to WebGPU. 

I'm looking at `ContextImplDML::CreateBufferImpl` and these imports seem feasible:

- A `WriteTo` Tensor could be imported as a GPUBuffer with `MAP_WRITE | COPY_SRC`.
- A `ReadFrom` Tensor could be imported as a GPUBuffer with  `MAP_READ | COPY_DST`.
- A default Tensor could be imported as a GPUBuffer with usages `COPY_SRC | COPY_DST | VERTEX | INDEX | INDIRECT | QUERY_RESOLVE |  STORAGE`.

I don't know that there's a usage model for all this, but on D3D12 they seem technically possible and the flexibility is interesting.

Reply:
Since we can't guarantee the underlying allocation parameters of an MLTensor when using the `WRITE` or `READ` hints, there should be a documented restriction on CreateTensor to not error when  `WEBGPU_INTEROP` is used with with `WRITE` or `READ` flags.

Also thinking longer term, if `GPUBufferUsages` were user configurable upon WebGPU import, the implementation would have some opportunity to optimize synchronization when during interop. (i.e. you don't need to synchronize when passing back to WebNN if WebGPU only accessed with read usages.)

Reply:
@a-sully 

In order to create a `GPUBuffer` from `MLTensor` by import, we need to map `MLTensorUsages` to compatible `GPUBufferUsage` (ex. `READ` => `MAP_READ`). Creating a `MLTensor` with `WEBGPU_INTEROP` will not be sufficient for import if we allow buffer usages to mismatch by making them [hints](https://github.com/webmachinelearning/webnn/pull/754#discussion_r1747825227). If `MLTensorUsages` flags were requirements for buffer creation then we could rely on importing a `MLTensor` buffer created with `READ`  as an equivalent `GPUBuffer` created with `STORAGE | MAP_READ`, for example (currently, `STORAGE` is implied).

Reply:
@bbernhar 

> If MLTensorUsages flags were requirements for buffer creation then we could rely on importing a MLTensor buffer created with READ as an equivalent GPUBuffer created with STORAGE | MAP_READ, for example (currently, STORAGE is implied).

FYI, in WebGPU `MAP_READ` can only be combined with `COPY_DEST`. Likewise, `MAP_WRITE` can only be combined with `COPY_SRC`[1]. `STORAGE` is not allowed on mappable buffers.

[1] https://www.w3.org/TR/webgpu/#buffer-usage

Reply:
That's interesting, @bjjones.

> STORAGE is not allowed on mappable buffers.

We probably need another `STORAGE`-like usage then as DML tensor buffers in WebNN are not necessarily assumed to be default buffers since they can remain readable in WebNN.

Reply:
> Is the ability to import an MLTensor into WebGPU still worthwhile?

For now, I think we can get away with ONLY allowing `GPUBufferUsageFlags.STORAGE` (implicit by WebNN) to be imported out [to WebGPU] upon creating the `GPUBuffer`. If the web developer needs `READ` or `WRITE` they must return the `GPUBuffer` back [to WebNN] in order to modify from script - only a WebGPU kernel can directly modify. Then WebGPU can safely combine-add other usages, `COPY_SRC` and/or `COPY_DST`, if required.

WebNN would also need to guarantee the `MLTensor` created with `READ` or `WRITE` will be made compatible (which may require a copy) with `GPUBufferUsageFlags.STORAGE` on creation when `WEBGPU_INTEROP` is used. Currently, WebNN's DML backend always allocates a `GPUBufferUsageFlags.STORAGE` compatible D3D12 buffer, regardless of memory/heap type, so it should be import-sharable (when possible).

Reply:
I concur with the above statement that we probably only need `STORAGE` usage initially. This will be possible to do within any path in the existing DML MLTensor implementation and should be very flexible for implementations on other backends.

I think `COPY_SRC` and `COPY_DST` could also be useful, but this does mean either:
   1. If using GPUBuffer as the imported type, the GPUBuffer implementation would need to support the GPUDevice.Copy* APIs. If an imported GPUBuffer needs to be backed by something different than it is now, this functionality would need implemented.
   2. If not using GPUBuffer as the imported type, we'd have to make new WebGPU APIs to make use of copy functionality. (i.e. `GPUDevice::CopyTextureToTensor`)

> Meanwhile, an MLMultiArray [may be backed by an IOSurface](https://developer.apple.com/documentation/coreml/mlmultiarray/init(pixelbuffer:shape:)) in some cases, which [can be imported to WebGPU](https://crsrc.org/c/gpu/command_buffer/service/shared_image/iosurface_image_backing.h).

Would it be possible to implement GPUBuffer backed with a 1D MTLTexture created from an IOSurface? You might be able to represent it in the shader using Metal Shading Language's `texture_buffer`.

Reply:
> For now, I think we can get away with ONLY allowing `GPUBufferUsageFlags.STORAGE`

Just to clarify, the current implementation in Chromium on DirectML uses `D3D12_HEAP_TYPE_READBACK` if creating an `MLTensor` with `MLTensorUsageFlags.READ`. If `READ | WEBGPU_INTEROP` is specified, then can a `GPUBuffer` with `GPUBufferUsageFlags.STORAGE` be created from this buffer? Or must we instead allocate the `MLTensor` using `D3D12_HEAP_TYPE_DEFAULT` in the first place?

Basically, can a `D3D12_HEAP_TYPE_READBACK` and `D3D12_HEAP_TYPE_UPLOAD` be used with `GPUBufferUsageFlags.STORAGE`?

If not, this is still relevant:

> then a user agent implementing a D3D-based backend must either:
> 
> * Allocate the `MLTensor` as a "default" buffer, requiring an additional staging buffer for `readBuffer()`, or
> * Allocate the `MLTensor` as a "readback" buffer and whip up a new GPU buffer for interop

---

> WebNN would also need to guarantee the `MLTensor` created with `READ` or `WRITE` will be made compatible (which may require a copy) with `GPUBufferUsageFlags.STORAGE` on creation when `WEBGPU_INTEROP` is used.

Well, either that or whip up a new GPU buffer for interop, as I described in the CoreML example above. That's an implementation detail, though.

> Would it be possible to implement GPUBuffer backed with a 1D MTLTexture created from an IOSurface? You might be able to represent it in the shader using Metal Shading Language's `texture_buffer`.

It seems so, yes https://developer.apple.com/documentation/metal/mtldevice/1433378-newtexturewithdescriptor

Reply:
> Basically, can a D3D12_HEAP_TYPE_READBACK and D3D12_HEAP_TYPE_UPLOAD be used with GPUBufferUsageFlags.STORAGE?

I believe so and `D3D12_HEAP_TYPE_DEFAULT` should be compatible with `GPUBufferUsageFlags.STORAGE` too. For WebNN's DML backend, `WEBGPU_INTEROP` would only be used for validation - the D3D heap types (and custom-equivalents used for UMA) only effect CPU access which `GPUBufferUsageFlags.STORAGE` forbids.

Reply:
Ack. Updated the explainer to be explicit that the imported buffer will have `GPUBufferUsageFlags.STORAGE`

Reply:
Resolving this thread now, but please feel free to un-resolve if there's more to discuss. Thanks!